pkg/kubectl/cmd/expose.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl"
+	"github.com/spf13/cobra"
+)
+
+func (f *Factory) NewCmdExposeService(out io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "expose <name> --port=<port> [--protocol=TCP|UDP] [--container-port=<number-or-name>] [--service-name=<name>] [--public-ip=<ip>] [--create-external-load-balancer]",
+		Short: "Take a replicated application and expose it as Kubernetes Service",
+		Long: `Take a replicated application and expose it as Kubernetes Service.
+		
+Looks up a ReplicationController named <name>, and uses the selector for that replication controller
+as the selector for a new Service which services on <port>
+
+Examples:
+$ kubectl expose nginx --port=80 --container-port=8000
+<creates a service for a replicated nginx, which serves on port 80 and connects to the containers on port 8000>
+
+$ kubectl expose streamer --port=4100 --protocol=udp --service-name=video-stream
+<create a service for a replicated streaming application on port 4100 balancing UDP traffic and is named 'video-stream'>
+`,
+		Run: func(cmd *cobra.Command, args []string) {
+			if len(args) != 1 {
+				usageError(cmd, "<name> is required for expose")
+			}
+
+			namespace := GetKubeNamespace(cmd)
+			client, err := f.Client(cmd)
+			checkErr(err)
+
+			rc, err := client.ReplicationControllers(namespace).Get(args[0])
+			checkErr(err)
+
+			generatorName := GetFlagString(cmd, "generator")
+			generator, found := kubectl.Generators[generatorName]
+			if !found {
+				usageError(cmd, fmt.Sprintf("Generator: %s not found.", generator))
+			}
+			if GetFlagInt(cmd, "port") < 1 {
+				usageError(cmd, "--port is required and must be a positive integer.")
+			}
+			names := generator.ParamNames()
+			params := kubectl.MakeParams(cmd, names)
+			if len(GetFlagString(cmd, "service-name")) == 0 {
+				params["name"] = args[0]
+			} else {
+				params["name"] = GetFlagString(cmd, "service-name")
+			}
+			params["labels"] = kubectl.MakeLabels(rc.Spec.Selector)
```

Comment:
I'd really like a clean separation of command-line parsing and business logic. This mixes the two by fetching the replication controller and extracting the labels in the body of the command. However, I don't have a better suggestion right now, since I don't think the generator should do the GET, and it seems like overkill to pass in the whole replication controller.